From 3e9fd572fe1a8bf03ca0fe6f0368b3cc03fd6a4e Mon Sep 17 00:00:00 2001
From: allen-bst <allen@bluestacks.com>
Date: Thu, 26 Oct 2023 11:05:36 +0800
Subject: [PATCH 039/224] [A13] Disabling location accuracy popup. Return false
 if gms is asking for network location provider and by default it is

---
 .../location/LocationManagerService.java      | 39 ++++++++++++++++++-
 1 file changed, 38 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/location/LocationManagerService.java b/services/core/java/com/android/server/location/LocationManagerService.java
index 3ff879661bda..8bd3f9bec7a4 100644
--- a/services/core/java/com/android/server/location/LocationManagerService.java
+++ b/services/core/java/com/android/server/location/LocationManagerService.java
@@ -47,6 +47,7 @@ import android.app.AppOpsManager;
 import android.app.PendingIntent;
 import android.app.compat.CompatChanges;
 import android.content.Context;
+import android.content.ContentResolver;
 import android.content.Intent;
 import android.content.pm.PackageManager;
 import android.location.Criteria;
@@ -89,8 +90,10 @@ import android.os.WorkSource;
 import android.os.WorkSource.WorkChain;
 import android.provider.Settings;
 import android.stats.location.LocationStatsEnums;
+import android.text.TextUtils;
 import android.util.ArrayMap;
 import android.util.ArraySet;
+import android.util.BstUtils;
 import android.util.IndentingPrintWriter;
 import android.util.Log;
 
@@ -232,7 +235,8 @@ public class LocationManagerService extends ILocationManager.Stub implements
     }
 
     public static final String TAG = "LocationManagerService";
-    public static final boolean D = Log.isLoggable(TAG, Log.DEBUG);
+    //public static final boolean D = Log.isLoggable(TAG, Log.DEBUG);
+    public static final boolean D = true;
 
     private static final String ATTRIBUTION_TAG = "LocationService";
 
@@ -1591,6 +1595,26 @@ public class LocationManagerService extends ILocationManager.Stub implements
 
         @Override
         public boolean isProviderEnabledForUser(@NonNull String provider, int userId) {
+
+            int uid = Binder.getCallingUid();
+            /*
+            * Changes to disable location accuracy popup. It popsup if network location provider is enabled.
+            * So returning false if gms is asking for the network location provider and by default it is disabled.
+            */
+            if ("network".equals(provider)) {
+                int pid = Binder.getCallingPid();
+                String packageName = BstUtils.getAppNameFromPid(pid);
+                if (D) Log.d(TAG, "isProviderEnabled pid=" + pid + ", uid=" + uid + ", packageName=" + packageName);
+                if (packageName != null && packageName.startsWith("com.google.android.gms")) {
+                    ContentResolver resolver = mContext.getContentResolver();
+                    boolean isLocProvEnabledForUser = isLocationProviderEnabledForUser(provider, userId);
+                    if (!isLocProvEnabledForUser) {
+                        if (D) Log.d(TAG, "Network is disabled for gms app so returning");
+                        return false;
+                    }
+                  }
+            }
+
             userId = ActivityManager.handleIncomingUser(Binder.getCallingPid(),
                     Binder.getCallingUid(), userId, false, false, "isProviderEnabledForUser", null);
 
@@ -1638,6 +1662,19 @@ public class LocationManagerService extends ILocationManager.Stub implements
             }
         }
 
+        private boolean isLocationProviderEnabledForUser(String provider, int userId) {
+            long identity = Binder.clearCallingIdentity();
+            try {
+                // Use system settings
+                ContentResolver cr = mContext.getContentResolver();
+                String allowedProviders = Settings.Secure.getStringForUser(
+                        cr, Settings.Secure.LOCATION_PROVIDERS_ALLOWED, userId);
+                return TextUtils.delimitedStringContains(allowedProviders, ',', provider);
+            } finally {
+                Binder.restoreCallingIdentity(identity);
+            }
+        }
+
         @Override
         public @Nullable LocationTime getGnssTimeMillis() {
             LocationProviderManager gpsManager = getLocationProviderManager(GPS_PROVIDER);
